File: test/workspace-changes.test.mjs
/**
 * 本轮改动文件投影：摘要 / 对比裁剪、坐标解析、历史内嵌与同轮取代。
 * 运行：node --test test/workspace-changes.test.mjs
 */
import { test } from "node:test"
import assert from "node:assert/strict"
import {
  MAX_DIFF_LINES,
  parseChangesCoordinates,
  projectChangesSummary,
  projectFileDiff,
  workspaceChangesService,
} from "../src/workspace-changes.js"
import { projectHistoryPage } from "../src/history.js"
import { pluginCapabilities } from "../src/protocol-caps.js"

const T = 1_700_000_000_000

function ev(seq, type, data) {
  return { event: { seq, type, time: T + seq, data } }
}

const summary = {
  turn: 3,
  cwd: "/Users/me/proj",
  files: [
    { path: "../up.txt", display: "../up.txt", added: 1, deleted: 0 },
    { path: "src/a.ts", display: "src/a.ts", added: 12, deleted: 3 },
    { path: "img.png", display: "img.png", added: 0, deleted: 0, binary: true },
    { path: "dist/big.map", display: "dist/big.map", added: 0, deleted: 0, oversized: true },
  ],
  total: 4,
  added: 13,
  deleted: 3,
  snapshot: { before: "aaa", after: "bbb" },
}

test("summary keeps order, flags and totals but drops cwd and snapshot ids", () => {
  const out = projectChangesSummary(summary)
  assert.deepEqual(out, {
    turn: 3,
    total: 4,
    added: 13,
    deleted: 3,
    files: [
      { path: "../up.txt", display: "../up.txt", added: 1, deleted: 0 },
      { path: "src/a.ts", display: "src/a.ts", added: 12, deleted: 3 },
      { path: "img.png", display: "img.png", added: 0, deleted: 0, binary: true },
      { path: "dist/big.map", display: "dist/big.map", added: 0, deleted: 0, oversized: true },
    ],
  })
  assert.equal("cwd" in out, false)
  assert.equal("snapshot" in out, false)
})

test("summary without files yields no card", () => {
  assert.equal(projectChangesSummary(undefined), undefined)
  assert.equal(projectChangesSummary({ ...summary, files: [], total: 0 }), undefined)
})

test("summary caps files but keeps the complete total", () => {
  const out = projectChangesSummary(summary, { maxFiles: 2 })
  assert.equal(out.files.length, 2)
  assert.equal(out.total, 4)
})

test("coordinates require non-negative integers", () => {
  const q = (s) => new URLSearchParams(s)
  assert.deepEqual(parseChangesCoordinates(q("seq=12")), { seq: 12 })
  assert.equal(parseChangesCoordinates(q("seq=-1")), null)
  assert.equal(parseChangesCoordinates(q("seq=1.5")), null)
  assert.equal(parseChangesCoordinates(q("")), null)
  assert.deepEqual(parseChangesCoordinates(q("seq=12&index=0"), { needIndex: true }), { seq: 12, index: 0 })
  assert.equal(parseChangesCoordinates(q("seq=12"), { needIndex: true }), null)
})

test("text diff passes through and truncates at the line cap", () => {
  const diff = {
    kind: "text",
    path: "src/a.ts",
    display: "src/a.ts",
    before: true,
    after: true,
    coarse: false,
    hunks: [
      { oldStart: 1, oldLines: 2, newStart: 1, newLines: 3, lines: [" a", "-b", "+c", "+d"] },
      { oldStart: 40, oldLines: 1, newStart: 41, newLines: 1, lines: ["-x", "+y"] },
    ],
  }
  const full = projectFileDiff(diff)
  assert.equal(full.hunks.length, 2)
  assert.equal(full.truncated, undefined)

  const cut = projectFileDiff(diff, { maxLines: 5 })
  assert.equal(cut.hunks.length, 2)
  assert.deepEqual(cut.hunks[1].lines, ["-x"])
  assert.equal(cut.hunks[1].oldStart, 40)
  assert.deepEqual(cut.truncated, { shownLines: 5, totalLines: 6 })

  const head = projectFileDiff(diff, { maxLines: 4 })
  assert.equal(head.hunks.length, 1)
  assert.deepEqual(head.truncated, { shownLines: 4, totalLines: 6 })
  assert.equal(MAX_DIFF_LINES, 5000)
})

test("text diff also stops at the character budget", () => {
  const long = "+" + "x".repeat(99)
  const diff = {
    kind: "text", path: "min.js", display: "min.js", before: true, after: true, coarse: false,
    hunks: [{ oldStart: 1, oldLines: 0, newStart: 1, newLines: 3, lines: [long, long, long] }],
  }
  const cut = projectFileDiff(diff, { maxChars: 250 })
  assert.deepEqual(cut.hunks[0].lines, [long, long])
  assert.deepEqual(cut.truncated, { shownLines: 2, totalLines: 3 })
})

test("binary / oversized diffs carry no lines", () => {
  assert.deepEqual(projectFileDiff({ kind: "binary", path: "img.png", display: "img.png" }), {
    kind: "binary", path: "img.png", display: "img.png",
  })
  assert.deepEqual(projectFileDiff({ kind: "oversized", path: "a", display: "a", hunks: [] }), {
    kind: "oversized", path: "a", display: "a",
  })
  assert.equal(projectFileDiff({ kind: "weird" }), undefined)
})

test("service lookup tolerates hosts without workspaceChanges", () => {
  assert.equal(workspaceChangesService(() => undefined), undefined)
  assert.equal(workspaceChangesService(() => { throw new Error("not provided") }), undefined)
  assert.equal(workspaceChangesService(() => ({ summary() {} })), undefined)
  const svc = { summary() {}, diff: async () => undefined }
  assert.equal(workspaceChangesService(() => svc), svc)
})

test("capabilities advertise changes only when the host serves them", () => {
  assert.equal(pluginCapabilities().files.changes, undefined)
  const caps = pluginCapabilities({ changes: true })
  assert.equal(caps.files.changes, true)
  assert.equal(caps.files.diff, true)
  assert.equal(caps.files.diffMaxLines, MAX_DIFF_LINES)
  assert.equal(caps.files.workspace, true)
})

test("history embeds the announced summary as a workspace_changes card", () => {
  const events = [
    ev(1, "user/message", { content: [{ type: "text", text: "改一下" }] }),
    ev(5, "workspace/changes", { turn: 3 }),
    ev(6, "turn/end", { turn: 3, reason: { kind: "completed" } }),
  ]
  const seen = []
  const page = projectHistoryPage({
    events,
    changesSummary: (seq) => { seen.push(seq); return summary },
  })
  assert.deepEqual(seen, [5])
  const card = page.messages.find((m) => m.role === "workspace_changes")
  assert.equal(card.id, "changes-5")
  assert.equal(card.seq, 5)
  assert.equal(card.turn, 3)
  assert.equal(card.changes.total, 4)
  assert.equal(card.changes.files[1].display, "src/a.ts")
})

test("history drops the card when the host no longer has the summary", () => {
  const events = [ev(5, "workspace/changes", { turn: 3 })]
  assert.equal(projectHistoryPage({ events, changesSummary: () => undefined }).messages.length, 0)
  assert.equal(projectHistoryPage({ events }).messages.length, 0)
  assert.equal(projectHistoryPage({ events, changesSummary: () => { throw new Error("boom") } }).messages.length, 0)
})

test("a later announcement for the same turn replaces the earlier card", () => {
  const later = { ...summary, files: [summary.files[1]], total: 1, added: 12, deleted: 3 }
  const events = [
    ev(5, "workspace/changes", { turn: 3 }),
    ev(9, "workspace/changes", { turn: 3 }),
  ]
  const page = projectHistoryPage({ events, changesSummary: (seq) => (seq === 5 ? summary : later) })
  const cards = page.messages.filter((m) => m.role === "workspace_changes")
  assert.equal(cards.length, 1)
  assert.equal(cards[0].id, "changes-9")
  assert.equal(cards[0].changes.total, 1)

  const emptied = projectHistoryPage({
    events,
    changesSummary: (seq) => (seq === 5 ? summary : { ...summary, files: [], total: 0 }),
  })
  assert.equal(emptied.messages.filter((m) => m.role === "workspace_changes").length, 0)
})

File: apps/android/app/src/test/java/dev/deeplinks/native/WorkspaceChangesTest.kt
package dev.deeplinks.native

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class WorkspaceChangesTest {

    private val summaryJson = """
        {"turn":3,"total":5,"added":13,"deleted":3,"files":[
          {"path":"../up.txt","display":"../up.txt","added":1,"deleted":0},
          {"path":"src/a.ts","display":"src/a.ts","added":12,"deleted":3},
          {"path":"img.png","display":"img.png","added":0,"deleted":0,"binary":true},
          {"path":"","display":"skip"},
          {"path":"dist/big.map","display":"dist/big.map","oversized":true}
        ]}
    """.trimIndent()

    private fun card(seq: Long, turn: Int, files: Int = 1) = workspaceChangesMessage(
        WorkspaceChangesSummary(
            seq = seq,
            turn = turn,
            total = files,
            added = 1,
            deleted = 0,
            files = (0 until files).map { ChangedFile("f$it", "f$it", added = 1) },
        ),
        time = seq,
    )

    private fun text(id: String, seq: Long, role: String = "assistant") =
        MobileMessage(id = id, role = role, text = id, seq = seq)

    @Test
    fun `summary parse keeps order and flags, drops blank paths, total never below files`() {
        val s = parseWorkspaceChanges(JSONObject(summaryJson), seq = 42)!!
        assertEquals(42L, s.seq)
        assertEquals(3, s.turn)
        assertEquals(listOf("../up.txt", "src/a.ts", "img.png", "dist/big.map"), s.files.map { it.display })
        assertTrue(s.files[2].binary)
        assertTrue(s.files[3].oversized)
        assertEquals(5, s.total)
        assertFalse(s.complete)
        assertEquals("a.ts", s.files[1].name)
        assertEquals("src", s.files[1].directory)
        assertEquals("", s.files[2].directory)

        val low = parseWorkspaceChanges(JSONObject("""{"total":0,"files":[{"path":"a"}]}"""), 1)!!
        assertEquals(1, low.total)
        assertTrue(low.complete)
    }

    @Test
    fun `summary without files yields no card`() {
        assertNull(parseWorkspaceChanges(JSONObject("""{"turn":1,"files":[]}"""), 1))
        assertNull(parseWorkspaceChanges(JSONObject("""{"turn":1}"""), 1))
    }

    @Test
    fun `diff parse covers text, truncation and degraded kinds`() {
        val text = parseWorkspaceFileDiff(
            JSONObject(
                """{"kind":"text","path":"a","display":"a","before":false,"after":true,"coarse":true,
                   "hunks":[{"oldStart":1,"oldLines":0,"newStart":1,"newLines":2,"lines":["+x","+y"]}],
                   "truncated":{"shownLines":2,"totalLines":9}}""",
            ),
        ) as WorkspaceFileDiff.Text
        assertTrue(text.truncated)
        assertEquals(listOf(DiffNote.CREATED, DiffNote.COARSE, DiffNote.TRUNCATED), diffNotes(text))
        assertTrue(parseWorkspaceFileDiff(JSONObject("""{"kind":"binary","path":"b"}""")) is WorkspaceFileDiff.Binary)
        assertTrue(parseWorkspaceFileDiff(JSONObject("""{"kind":"oversized","path":"c"}""")) is WorkspaceFileDiff.Oversized)
        assertNull(parseWorkspaceFileDiff(JSONObject("""{"kind":"weird"}""")))

        val same = WorkspaceFileDiff.Text("a", "a", before = true, after = false, coarse = false, hunks = emptyList())
        assertEquals(listOf(DiffNote.DELETED, DiffNote.UNCHANGED), diffNotes(same))
    }

    @Test
    fun `diff rows number both sides from hunk starts`() {
        val rows = diffRows(
            listOf(DiffHunk(oldStart = 40, oldLines = 3, newStart = 41, newLines = 3, lines = listOf(" a", "-b", "+c", " d", ""))),
        )
        assertEquals(DiffRow.Kind.HUNK, rows[0].kind)
        assertEquals("@@ -40,3 +41,3 @@", rows[0].text)
        assertEquals(DiffRow(DiffRow.Kind.CONTEXT, 40, 41, "a"), rows[1])
        assertEquals(DiffRow(DiffRow.Kind.DELETE, 41, null, "b"), rows[2])
        assertEquals(DiffRow(DiffRow.Kind.ADD, null, 42, "c"), rows[3])
        assertEquals(DiffRow(DiffRow.Kind.CONTEXT, 42, 43, "d"), rows[4])
        // 空行（缺前缀）当作上下文，不崩
        assertEquals(DiffRow(DiffRow.Kind.CONTEXT, 43, 44, ""), rows[5])
    }

    @Test
    fun `later announcement for the same turn replaces the earlier card`() {
        val list = listOf(text("u", 1, "user"), card(5, turn = 3), text("m", 6), card(9, turn = 3, files = 2), card(12, turn = 4))
        val out = coalesceWorkspaceChanges(list)
        assertEquals(listOf("u", "m", "changes-9", "changes-12"), out.map { it.id })
        assertEquals(listOf(12L, 9L), sessionChangeSummaries(list).map { it.seq })
        // 没有改动卡的列表原样返回（同一实例，不做多余拷贝）
        val plain = listOf(text("a", 1))
        assertTrue(coalesceWorkspaceChanges(plain) === plain)
    }

    @Test
    fun `history pages merge coalesces cards across the page boundary`() {
        val older = listOf(text("u1", 1, "user"), card(5, turn = 1))
        val tail = listOf(card(8, turn = 1), text("u2", 10, "user"))
        assertEquals(listOf("u1", "changes-8", "u2"), mergeHistoryPages(older, tail).map { it.id })
    }

    @Test
    fun `live card survives a refresh that has not reached its announcement`() {
        val live = listOf(text("u", 1, "user"), card(7, turn = 1))
        val fresh = listOf(text("u", 1, "user"), text("m", 5))
        assertEquals(listOf("u", "m", "changes-7"), mergeHistoryWithLive(fresh, live).map { it.id })
    }

    @Test
    fun `live card is dropped once history covers it without a card`() {
        val live = listOf(text("u", 1, "user"), card(7, turn = 1))
        // history 已推进到 seq 9 仍没有这张卡：同轮被取代为空或 Host 已取不到摘要
        val fresh = listOf(text("u", 1, "user"), text("m", 9))
        assertEquals(listOf("u", "m"), mergeHistoryWithLive(fresh, live).map { it.id })
    }

    @Test
    fun `history card with same id wins over the live copy`() {
        val live = listOf(card(7, turn = 1, files = 1))
        val fresh = listOf(card(7, turn = 1, files = 3))
        val merged = mergeHistoryWithLive(fresh, live)
        assertEquals(1, merged.size)
        assertEquals(3, merged[0].changes?.files?.size)
    }

    @Test
    fun `raw fallback never swallows the changes card`() {
        assertFalse(isRawFallback(card(1, turn = 1)))
    }

    @Test
    fun `panel is full screen on phones and keeps 400dp for the main column on tablets`() {
        assertEquals(412f, changesPanelWidthDp(412f))
        assertEquals(599f, changesPanelWidthDp(599f))
        // 840dp：45% = 378，主区剩 462 ≥ 400
        assertEquals(378f, changesPanelWidthDp(840f), 0.01f)
        // 1280dp：45% = 576
        assertEquals(576f, changesPanelWidthDp(1280f), 0.01f)
        // 700dp：45% = 315 → 至少 360；主区只剩 340 < 400 时仍保 360 的可用面板
        assertEquals(360f, changesPanelWidthDp(700f), 0.01f)
    }

    @Test
    fun `release settles by fling first, then by distance`() {
        assertTrue(settleChangesPanelOpen(0.1f, 2_000f, 1_000f))
        assertFalse(settleChangesPanelOpen(0.9f, -2_000f, 1_000f))
        assertTrue(settleChangesPanelOpen(0.5f, 0f, 1_000f))
        assertFalse(settleChangesPanelOpen(0.2f, 300f, 1_000f))
    }
}
